### PR TITLE
Foundation fixes for 32-bit systems and more

### PR DESCRIPTION
**Pull request by hpux735**
This is a collection of fixes that address some edge cases relating to 32-bit systems, differing alignments, uninitialized locales, and incomplete paths.

### CODE REVIEW

**Comment by hpux735** [comment]
cc @tienex

File: CoreFoundation/NumberDate.subproj/CFNumber.c [hunk 1]
```diff
@@ -1202,7 +1202,7 @@ CFNumberRef CFNumberCreate(CFAllocatorRef allocator, CFNumberType type, const vo
 	}
     }
 
-    CFIndex size = 8 + ((!__CFNumberTypeTable[type].floatBit && __CFNumberTypeTable[type].storageBit) ? 8 : 0);
+    CFIndex size = (sizeof(struct __CFNumber) - sizeof(CFRuntimeBase)) + ((!__CFNumberTypeTable[type].floatBit && __CFNumberTypeTable[type].storageBit) ? 8 : 0);
```
Comment (phausler) [inline CoreFoundation/NumberDate.subproj/CFNumber.c]: Interesting, this seems to work just fine on 32 bit linked processes; I will have to do some digging to verify that...
The change listed here actually looks more correct.

File: Foundation/NSGeometry.swift [hunk 2]
```diff
@@ -33,7 +33,11 @@ public struct CGFloat {
     public var native: NativeType
     
     private var hash: Int {
+#if arch(i386) || arch(arm)
```
Comment (phausler) [inline Foundation/NSGeometry.swift]: I wish we had a #if 32bit or some sort of equiv; ala the discussion on swift-evolution.

Comment (hpux735) [inline Foundation/NSGeometry.swift]: YES!!  I agree.  I'd also love an endianness flag. :)

Comment (tienex) [inline Foundation/NSGeometry.swift]: I have my tree where I have done endian([big|little]) and intsize(nn) but reading the discussion I decided to postpone the submission. :)

File: CoreFoundation/NumberDate.subproj/CFNumber.c (hunk 1 above)
Comment (tienex) [inline CoreFoundation/NumberDate.subproj/CFNumber.c]: On Linux/arm the alignment is different, the _pad is found at 24, while CFRuntimeBase is 20 bytes long leading to an awful heap overflow.

**Comment by hpux735** [comment]
@phausler I'm just checking up on a few pending PRs that I have...  Is there anything on this one that you need me to fix?  (sorry to be a pest)

**Comment by phausler** [comment]
everything checks out

**Comment by hpux735** [comment]
Thanks!

**Pull request merged**